flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/dsv2/eventtime/StatisticNewsClickNumber.java

```diff
@@ -0,0 +1,278 @@
+        }
+
+        env.execute("StatisticNewsClickNumberExample");
+    }
+
+    /**
+     * This process function will consume {@link NewsEvent} and count the number of clicks within 1
+     * minute, 5 minutes, 10 minutes, 30 minutes and 1 hour of the news releasing and send the
+     * results {@link NewsClickNumber} to the output.
+     *
+     * <p>To achieve the goal, we will register a series of timers for the news, which will be
+     * triggered at the time of the news's release time + 1 minute/5 minutes/10 minutes/30 minutes/1
+     * hour, and record a list of the click times of each news. In the timer callback {@code
+     * onEventTimer}, we will count the number of clicks between the news release time and the timer
+     * trigger timer and send the result to the output.
+     */
+    public static class CountNewsClickNumberProcessFunction
+            implements OneInputEventTimeStreamProcessFunction<NewsEvent, NewsClickNumber> {
+
+        private EventTimeManager eventTimeManager;
+
+        // news id to release time
+        private final Map<Long, Long> releaseTimeOfNews = new HashMap<>();
+
+        // news id to click time list
+        private final Map<Long, List<Long>> clickTimeListOfNews = new HashMap<>();
+
+        @Override
+        public void initEventTimeProcessFunction(EventTimeManager eventTimeManager) {
+            this.eventTimeManager = eventTimeManager;
+        }
+
+        @Override
+        public void processRecord(
+                NewsEvent record,
+                Collector<NewsClickNumber> output,
+                PartitionedContext<NewsClickNumber> ctx)
+                throws Exception {
+            if (record.type == NewsEventType.RELEASE) {
+                // for the news release event, record the release time and register timers
+                long releaseTime = record.timestamp;
+                releaseTimeOfNews.put(record.newsId, releaseTime);
+                for (Duration targetTime : TIMES_AFTER_NEWS_RELEASE) {
+                    eventTimeManager.registerTimer(releaseTime + targetTime.toMillis());
+                }
+            } else {
+                // for the news click event, record the click time
+                clickTimeListOfNews
+                        .computeIfAbsent(record.newsId, k -> new ArrayList<>())
+                        .add(record.timestamp);
+            }
+        }
+
+        @Override
+        public void onEventTimer(
+                long timestamp,
+                Collector<NewsClickNumber> output,
+                PartitionedContext<NewsClickNumber> ctx) {
+            // get the news that the current event timer belongs to
+            long newsId = ctx.getStateManager().getCurrentKey();
+
+            // calculate the difference between the current time and the news release time
+            Duration diffTime = Duration.ofMillis(timestamp - releaseTimeOfNews.get(newsId));
```

Comment:
Need normalization.

Reply:
Maybe for simplicity, we just need one counting period.
1. We maintain a counter for each news id. When receiving new clicks, we check whether it falls into the desired time period, and increase the counter only if it does.
2. When the event timer triggers, we output the counter and clear the states.
3. We should also consider the case where publishing event is received after clicking events. In that case, we need to cache the early arrived clicking events somewhere, and handle and clear them once the publishing event is received.
4. We should also consider the case where publishing event is never received, and have a timer to check and clear the cached events.